pkgs/desktops/deepin/default.nix

```diff
@@ -0,0 +1,52 @@
+{ lib, pkgs, libsForQt5 }:
+let
+  functions = with pkgs.lib; rec {
+    getPatchFrom' = commonRp:
+      let
+        rpstr = a: b: " --replace \"${a}\" \"${b}\"";
+        rpstrL = l: if length l == 2 then rpstr (head l) (last l) else (throw "input must be a 2-tuple");
+        rpfile = filePath: replaceLists:
+          "substituteInPlace ${filePath}" + concatMapStrings rpstrL replaceLists;
+      in
+      x: pipe x [
+        (x: mapAttrs (name: value: value ++ commonRp) x)
+        (x: mapAttrsToList (name: value: rpfile name value) x)
+        (concatStringsSep "\n")
+        (s: s + "\n")
+      ];
+
+    getPatchFrom = getPatchFrom' [ ];
+    getUsrPatchFrom = getPatchFrom' [ [ "/usr" "$out" ] ];
+
+    replaceAll = x: y: ''
+      echo Replacing "${x}" to "${y}":
+      for file in $(grep -rl "${x}")
+      do
+        echo -- $file
+        substituteInPlace $file \
+          --replace "${x}" "${y}"
+      done
+    '';
+  };
```

Comment:
These functions seem to be unused?

Reply:
There are too much  hardcode path in dde,  library in this pr does need these functions，but some of DDE core components need this function to generate the patch：

https://github.com/linuxdeepin/dde-nixos/blob/main/packages/core/dde-file-manager/default.nix
https://github.com/linuxdeepin/dde-nixos/blob/main/packages/go-package/dde-daemon/default.nix
https://github.com/linuxdeepin/dde-nixos/blob/main/packages/core/dde-session-shell/default.nix
https://github.com/linuxdeepin/dde-nixos/blob/main/packages/core/dde-session-ui/default.nix